### ISSUE TITLE
kvserver: 22.2 range descriptors can prevent 22.1 range config changes

### ISSUE THREAD

**Issue by erikgrinaker** [cockroachdb/cockroach#94834]
In #85120, released in 22.2, we changed `RangeDescriptor.StickyBit` and `ReplicaDescriptor.Type` to be non-nullable (these use Protobuf 2 which supports optional scalar fields). This was believed to be safe, because 22.1 considers a `nil` value the same as the zero value, e.g. via [`GetStickyBit()`](https://github.com/cockroachdb/cockroach/blob/1e03d0c43eed56028c1b6d9b6b17ea25304c9ab6/pkg/roachpb/metadata.go#L338-L344) and [`GetType()`](https://github.com/cockroachdb/cockroach/blob/1e03d0c43eed56028c1b6d9b6b17ea25304c9ab6/pkg/roachpb/metadata.go#L452-L458), and #38302 ensured we did not rely on range descriptors having a stable wire encoding.

Unfortunately, `RangeDescriptor.Equal()` and `ReplicaDescriptor.Equal()` considered a `nil` value to be different from a zero value, unlike the rest of the code. This was ok in 22.1 because we [always normalized the zero value to `nil`](https://github.com/cockroachdb/cockroach/blob/1e03d0c43eed56028c1b6d9b6b17ea25304c9ab6/pkg/roachpb/metadata.go#L262-L267). However, these comparisons will fail on 22.1 when comparing the 22.2 representation of a range descriptor (with an explicit zero value) to the 22.1 representation (with a nil value).

Notably, we do such a comparison [when processing range config changes](https://github.com/cockroachdb/cockroach/blob/d4a2c743e02c82a903fdbac782f1d5df5e8c33a1/pkg/kv/kvserver/replica_command.go#L2243) (e.g. moving a replica, up/downreplication, splits, merges, etc), to ensure the caller's view of the range descriptor matches the canonical range descriptor stored in KV. Most of the time this works fine, because in most cases the caller's range descriptor originates on the same node that's processing the config change, but there are a few cases where it doesn't, in which case the config change will fail with an error like the following:

```
change replicas of r41 failed: descriptor changed: [expected] r41:/Table/4{5-6} [(n1,s1):1, (n3,s3):2, (n4,s4):4, next=5, gen=8] != [actual] r41:/Table/4{5-6} [(n1,s1):1, (n3,s3):2, (n4,s4):4, next=5, gen=8]
```

Notice how the range descriptors appear to be equal, but the comparison still fails.

There are two main scenarios where this can happen:

1. In a mixed 22.2/22.1 cluster, or following a downgrade from 22.2 to 22.1, a 22.1 leaseholder's in-memory representation of the range descriptor can diverge from the canonical range descriptor stored in KV, in which case all replicate/split/merge queue operations (i.e. all config changes carried out by CRDB itself) for that range will fail.

    This typically happens when a 22.1 leaseholder has previously moved its replica to a different node and given its lease to a 22.2 node who committed the replica move, and the lease is then later transferred back to a different 22.1 replica. The original 22.1 leaseholder will [write `Type:nil`](https://github.com/cockroachdb/cockroach/blob/d4a2c743e02c82a903fdbac782f1d5df5e8c33a1/pkg/kv/kvserver/replica_command.go#L2315) to the canonical range descriptor in KV, but when the config change transaction is committed by the new 22.2 leaseholder the range descriptor is [propagated with `Type:VOTER_FULL` via Raft](https://github.com/cockroachdb/cockroach/blob/6133d88a46a984704a700d4e7b251f82ee212231/pkg/kv/kvserver/batcheval/cmd_end_transaction.go#L1260-L1262) which is then stored in the in-memory state of all replicas.

    Note that this can only happen in clusters with more nodes than the replication factor, since there's otherwise nowhere to move a replica to (e.g. won't happen in 3-node clusters with RF=3 or 5-node clusters with RF=5).

2. In a mixed-version cluster consisting of 22.2 and 22.1 nodes, `AdminChangeReplicas` requests from a 22.2 node to a 22.1 leaseholder will fail, because the request [carries](https://github.com/cockroachdb/cockroach/blob/5fb16c8e1bcb6f9dff7c51cb5d59a2c45dd4751f/pkg/roachpb/api.proto#L857) the calling node's view of the range descriptor. This typically happens when running e.g. `ALTER RANGE RELOCATE` on a 22.2 SQL gateway, since most internal CRDB requests typically originate on the local node. Running the request from a 22.1 node will work with both 22.1 and 22.2 leaseholders. `ALTER TABLE SPLIT` is not affected, unless via 1 above, since it uses the leaseholder's view of the descriptor.

In both cases, the problem can be resolved either by upgrading all nodes to 22.2 (finalizing the upgrade is not necessary), downgrading to 22.1, and/or restarting any affected nodes to reset the in-memory range descriptor state to the canonical KV representations.

Jira issue: CRDB-23172

**Comment by erikgrinaker** [comment]
This also applies to `RangeDescriptor.StickyBit`.

**Comment by erikgrinaker** [comment]
Draft PR up in #94834. Will need some additional scrutiny.

**Comment by erikgrinaker** [comment]
There is a secondary problem here in that the replica's in-memory representation of the range descriptor may diverge from the KV representation, possibly because of normalizations like this one:

https://github.com/cockroachdb/cockroach/blob/d4a2c743e02c82a903fdbac782f1d5df5e8c33a1/pkg/kv/kvserver/replica_command.go#L162-L174

The replicate queue uses the replica's in-memory descriptor when making replication changes, so if the in-memory representation diverges from the KV representation then it will fail the comparison as well. The fix in #94841 should address that as well. I'll need to confirm this and audit the code more closely.

**Comment by erikgrinaker** [comment]
We've confirmed two known problems with this, I've updated the issue description to reflect recent findings. We have fixes in flight both for 22.1 (for compatibility with 22.2.0-22.2.2) and 22.2 (for compatibility with all 22.1 versions).

**Comment by erikgrinaker** [comment]
I've audited all `Equal()` calls in 22.1:

### RangeDescriptor.Equal()

* `rangecache.compareEntryDescs()`: unaffected, will fall through to further comparison logic.
    https://github.com/cockroachdb/cockroach/blob/705ec0e36c4f0e7c9eb8f2b1b6ebe925f674f2a0/pkg/kv/kvclient/rangecache/range_cache.go#L1246-L1248

* `kvserver.checkDescsEqual()`: the cause of these bugs, may affect all config changes including splits/merges.
    https://github.com/cockroachdb/cockroach/blob/d4a2c743e02c82a903fdbac782f1d5df5e8c33a1/pkg/kv/kvserver/replica_command.go#L2692-L2696

* `replicaGCQueue.process()`: affected but insignificant, may prevent replica GC on in-memory descriptor divergence in rare cases.
    https://github.com/cockroachdb/cockroach/blob/9f47a214c41e14af661a07600c66ddcdbf525bbc/pkg/kv/kvserver/replica_gc_queue.go#L346-L353

* `statusServer.RaftDebug()`: affected but insignificant, may cause errors with Raft debug endpoints.
    https://github.com/cockroachdb/cockroach/blob/4bbe33efc8aaaaf2bf952e3a126d86465c07027a/pkg/server/status.go#L1826-L1831

* `replica.assertStateRaftMuLockedReplicaMuRLocked()`: unaffected (thankfully 😅), because the disk state is populated with the replica's in-memory state before comparison.

    https://github.com/cockroachdb/cockroach/blob/bd3c083f3987b9066bec676f5065174be54f5d3f/pkg/kv/kvserver/replica.go#L1304
    https://github.com/cockroachdb/cockroach/blob/bd3c083f3987b9066bec676f5065174be54f5d3f/pkg/kv/kvserver/replica.go#L1295

### ReplicaDescriptor.Equal()

* `CacheEntry.updateLease()`: affected but believed to be insignificant, may cause lease cache thrashing when 22.1 nodes receive NLHE from 22.2 nodes, but should stabilize once they find the actual leaseholder.
https://github.com/cockroachdb/cockroach/blob/705ec0e36c4f0e7c9eb8f2b1b6ebe925f674f2a0/pkg/kv/kvclient/rangecache/range_cache.go#L1333-L1335

* `replica.assertStateRaftMuLockedReplicaMuRLocked()`: affected via `ReplicaState.Lease`, but believed to be insignificant since this is synced to disk.
    https://github.com/cockroachdb/cockroach/blob/bd3c083f3987b9066bec676f5065174be54f5d3f/pkg/kv/kvserver/replica.go#L1304

**Comment by erikgrinaker** [comment]
Resolved by #94888 and #94841, which will be included in 22.2.3 (January 23rd) and 22.1.14 (February 6th). Either fix is sufficient by itself.
